File: projeto_construcao/src/main/java/br/com/projeto/domain/service/UserService.java
package br.com.projeto.domain.service;

import org.directwebremoting.annotations.RemoteMethod;
import org.directwebremoting.annotations.RemoteProxy;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Sort.Direction;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.security.access.prepost.PreAuthorize;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import br.com.projeto.domain.entity.User;
import br.com.projeto.domain.repository.IUserRepository;
import br.com.projeto.infrastructure.Mailer;

/**
 * 
 * @author André
 * @category Service
 * 
 */
@RemoteProxy
@Service("userService")
@Transactional
public class UserService
{
	/*-------------------------------------------------------------------
	 *				 		     ATTRIBUTES
	 *-------------------------------------------------------------------*/
	//Repository
	/**
	 * 
	 */
	@Autowired
	private IUserRepository userRepository;
	
	/**
	 * 
	 */
	@Autowired
	private Mailer mailer;
	
	/*-------------------------------------------------------------------
	 *				 		     SERVICES
	 *-------------------------------------------------------------------*/
	/**
	 * 
	 * @param user
	 * @return
	 */
	@RemoteMethod
	@PreAuthorize("hasRole('ROLE_ADMIN')")
	public ResponseEntity<String>  insertUser(User user)
	{		
		if ( user == null )
		{
			return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR).body("Usuário nulo");
		}
		if ( (userRepository.findByEmail(user.getEmail(), new Long(0)) != null ) )
		{
			return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR).body("Email já cadastrado");
		}
		if ( !user.passwordIsValid() )
		{
			return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR).body("Senhas não conferem");
		}

		mailer.send(user); 
		
		String hash = new BCryptPasswordEncoder().encode(user.getPassword()); 
		user.setPassword(hash);
		user.setActive(true);
	    userRepository.save(user);
	    return ResponseEntity.status(HttpStatus.ACCEPTED).body("Usuário salvo com sucesso!");
	}
	/**
	 * 
	 * @param id
	 * @return
	 */
	@Transactional(readOnly = true)
	@RemoteMethod
	public User findUserById(Long id) 
	{
		return userRepository.findOne(id);
	}	
	/**
	 * 
	 * @return
	 */
	@Transactional(readOnly = true)
	@RemoteMethod
	public String helloWorld() 
	{
		return "Olá Mundo";
	}
	
	/**
	 * 
	 * @return
	 */
	@Transactional(readOnly = true)
	@RemoteMethod
	public User getCurrent()
	{	
		User userCurrent = (User) SecurityContextHolder.getContext().getAuthentication().getPrincipal();
		return userCurrent;
	}
	
	/**
	 * 
	 * @param user
	 * @return 
	 */
	@RemoteMethod
	@PreAuthorize("hasRole('ROLE_ADMIN')")
	public ResponseEntity<String> updateUsertoActivate(User user) 
	{
		if ( user == null )
		{
			return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR).body("Usuário nulo");
		}
		user.setActive(true);
		userRepository.saveAndFlush(user);
		return ResponseEntity.status(HttpStatus.ACCEPTED).body("Usuário ativado com sucesso!");
	}
	/**
	 * 
	 * @param user
	 * @return 
	 * @return
	 */
	@RemoteMethod
	@PreAuthorize("hasRole('ROLE_ADMIN')")
	public ResponseEntity<String> updateUsertoDeactivate(User user) 
	{
		User userCurrent = (User) SecurityContextHolder.getContext().getAuthentication().getPrincipal();
		if ( user == null )
		{
			return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR).body("Usuário nulo");
		}
		if ( ( user.getId() == 1 ) || ( user.getId() == userCurrent.getId() ))
		{
			throw new IllegalArgumentException("O usuário não pode ser desativado.");
		}
		user.setActive(false);
		userRepository.saveAndFlush(user);
		return ResponseEntity.status(HttpStatus.ACCEPTED).body("Usuário desativado com sucesso!");
	}
	/**
	 * 
	 * @param user
	 * @return 
	 */
	@RemoteMethod
	@PreAuthorize("hasRole('ROLE_ADMIN')")
	public ResponseEntity<String> updateUser(User user) 
	{
		if ( user == null )
		{
			return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR).body("Usuário nulo");
		}
		User currentUser = userRepository.findOne(user.getId());
		if ( (userRepository.findByEmail(user.getEmail(), user.getId()) != null ) )
		{
			return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR).body("Email já cadastrado");
		}
		user.setPassword(currentUser.getPassword());
		userRepository.saveAndFlush(user);
		return ResponseEntity.status(HttpStatus.ACCEPTED).body("Usuário salvo com sucesso!");	 
	}
	
	/**
	 * 
	 * @param user
	 * @return
	 */
	@RemoteMethod
	@PreAuthorize("hasRole('ROLE_ADMIN')")
	public ResponseEntity<String> updateUserToPassword(User user) 
	{
		if ( user == null )
		{
			return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR).body("Usuário nulo");
		}
		if ( !user.passwordIsValid() )
		{
			 return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR).body("Senhas não conferem");
		}
		String hash = new BCryptPasswordEncoder().encode(user.getPassword());
		user.setPassword(hash);
		userRepository.saveAndFlush(user);	 
		return ResponseEntity.status(HttpStatus.ACCEPTED).body("Senha atualizada com sucesso!");
	}

	/**
	 * 
	 * @param page
	 * @param size
	 * @param property
	 * @param order
	 * @param filter
	 * @return
	 */
	@Transactional(readOnly = true)
	public Page<User> listUsersByFilters(int page, int size, String property, String order, String filter) 
	{
		Direction asc;
		if ( filter.compareToIgnoreCase("null") == 0 )
		{
			filter = "";
		}
		if (order.equals("ASC"))
		{
			asc = Direction.ASC;
		}
		else
		{
			asc = Direction.DESC;
		}
		PageRequest pageable = new PageRequest(page, size, asc, property);
		return userRepository.listUsersByFilters(filter.toLowerCase(), pageable);
	}
}

File: Angular-material/angular-covalent/src/app/service/user.service.ts
import { PageRequest } from './../model/PageRequest';
import { User } from './../model/User';


import { Observable } from 'rxjs/Observable';
import { Http, Headers, Response } from '@angular/http';
import { Injectable } from '@angular/core';

/**
 * 
 */
@Injectable()
export class UserService 
{
 
  /*-------------------------------------------------------------------
	 * 		 					ATTRIBUTES
	 *-------------------------------------------------------------------*/
  /**
   * 
   */
  headers: Headers;
  /**
   * 
   */
  url: String = '/projeto/api/user/';

 
  /*-------------------------------------------------------------------
	 * 		 					CONSTRUCTOR
	 *-------------------------------------------------------------------*/
  constructor(public http: Http) 
  { 
     this.headers = new Headers();
     this.headers.append('Content-Type', 'application/json');
  }

  /*-------------------------------------------------------------------
	 *				 		     SERVICES
	 *-------------------------------------------------------------------*/

  /**
   * 
   * @param user 
   */
  insertUser(user): Observable<Response>
  {
    if(user.id != undefined )
    {
       return this.http.put(this.url + 'updateUser', JSON.stringify(user), { headers: this.headers });
    }
    else 
    {
       return this.http.post(this.url + 'insertUser', JSON.stringify(user), { headers: this.headers });
    }
  }
  /**
   * 
   * @param user 
   */
  updateUsertoActivate(user): Observable<Response>
  {
    return this.http.patch(this.url + 'updateUsertoActivate/' + user.id, JSON.stringify(user), { headers: this.headers });
  }
  /**
   * 
   * @param user 
   */
  updateUsertoDeactivate(user): Observable<Response>
  {
    return this.http.patch(this.url + 'updateUsertoDeactivate/' + user.id, JSON.stringify(user), { headers: this.headers });
  }
  /**
   * 
   * @param id 
   */
  findUserbyId(id): Observable<User>
  {
    return this.http.get(this.url + 'findUserById/' + id).map(res => res.json());;
  }
  /**
   * 
   */
  getCurrentUser(): Observable<User>
  {
    return this.http.get(this.url + 'getCurrentUser').map(res => res.json());;
  } 
  /**
   * 
   * @param user 
   */
  updateUserToPassword(user): Observable<Response>
  {
    return this.http.put(this.url + 'updateUserToPassword', JSON.stringify(user), { headers: this.headers });
  }
  /**
   * 
   * @param page 
   * @param size 
   * @param property 
   * @param order 
   * @param textSearch 
   */
  listUsersByFilters(page: number, size: number, property: String, order: String, textSearch: String): Observable<PageRequest>
  {
    return this.http.get(this.url + 'listUsersByFilters/'+ page + '/' + size + '/' + property + '/' + order + '/' + textSearch).map(res => res.json());
  } 
}